rad/src/lib.rs

```diff
@@ -349,40 +357,118 @@ fn add_http_headers(
 pub async fn run_retrieval_report(
     retrieve: &RADRetrieve,
     settings: RadonScriptExecutionSettings,
-    active_wips: ActiveWips,
+    active_wips: &ActiveWips,
+    client: Option<WitnetHttpClient>,
 ) -> Result<RadonReport<RadonTypes>> {
     let context = &mut ReportContext::from_stage(Stage::Retrieval(RetrievalMetadata::default()));
-    context.set_active_wips(active_wips);
+    context.set_active_wips(active_wips.clone());
 
     match retrieve.kind {
-        RADType::HttpGet => http_response(retrieve, context, settings).await,
+        RADType::HttpGet => http_response(retrieve, context, settings, client).await,
         RADType::Rng => rng_response(context, settings).await,
-        RADType::HttpPost => http_response(retrieve, context, settings).await,
+        RADType::HttpPost => http_response(retrieve, context, settings, client).await,
         _ => Err(RadError::UnknownRetrieval),
     }
 }
 
 /// Run retrieval stage of a data request, return `Result<RadonTypes>`.
-pub async fn run_retrieval(retrieve: &RADRetrieve, active_wips: ActiveWips) -> Result<RadonTypes> {
+pub async fn run_retrieval(retrieve: &RADRetrieve, active_wips: &ActiveWips) -> Result<RadonTypes> {
     // Disable all execution tracing features, as this is the best-effort version of this method
     run_retrieval_report(
         retrieve,
         RadonScriptExecutionSettings::disable_all(),
         active_wips,
+        None,
     )
     .await
     .map(RadonReport::into_inner)
 }
 
+/// Run retrieval using multiple transports, and only produce a positive result if the retrieved
+/// values pass the filter function from the tally stage.
+///
+/// The idea behind this is to avoid producing commitments for data requests with sources that act
+/// in an inconsistent way, i.e. they return very different values when queried through different
+/// HTTP transports at once.
+pub async fn run_paranoid_retrieval(
+    retrieve: &RADRetrieve,
+    tally: &RADTally,
```

Comment:
It looks strange that this uses the retrieve stage and the tally stage, but does not use the aggregation stage. Maybe it would be better to run all retrievals + aggregation for each proxy, and then apply the paranoid logic to the vector of aggregated results. Then this PR would not be detecting "inconsistent sources" but "inconsistent requests", but I guess that's fine?

Reply:
I had the same doubt when I was revisiting this.

Reply:
This was utterly wrong.

I was using the tally executor for a good reason (it accumulates much more metadata that we need for evaluating consistency), but I was wrong passing the tally instead of the aggregation.

Now I'm taking the aggregation and synthetically turning it into a tally.

In the future, if we think that accumulating the metadata for aggregations is useful for debugging or whatever, we can refactor this part and simply use the aggregator executor.